confmap/resolver.go

```diff
@@ -187,3 +188,82 @@ func (mr *Resolver) closeIfNeeded(ctx context.Context) error {
 	}
 	return err
 }
+
+func (mr *Resolver) expandValueRecursively(ctx context.Context, value interface{}) (interface{}, error) {
+	for i := 0; i < 100; i++ {
+		val, changed, err := mr.expandValue(ctx, value)
+		if err != nil {
+			return nil, err
+		}
+		if !changed {
+			return val, nil
+		}
+		value = val
+	}
+	return nil, errors.New("too many recursive expansions")
+}
+
+func (mr *Resolver) expandValue(ctx context.Context, value interface{}) (interface{}, bool, error) {
+	switch v := value.(type) {
+	case string:
+		// If it doesn't have the format "${scheme:opaque}" no need to expand.
```

Comment:
Something like `${HOST_IP}:4317` does not get expanded then? Why don't we use `os.Expand` like we are using on the expand provider?

Reply:
Because `os.Expand` only supports returning strings (not a partial sub-map).

Reply:
Right, but the problem of dealing with `${HOST_IP}:4317` still remains then? That would only be supported by environment variables?

Reply:
Do we want to support this? What is a good use-case?

Reply:
Here are some examples on contrib using this with environment variables:
- [sapm exporter](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/8782bd5d6b6bf44ae3d214b5c8c0d56e8293e744/exporter/sapmexporter/examples/signalfx-k8s.yaml#L71 )
- [kubeletstats receiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/kubeletstatsreceiver#service-account-authentication-example) (see `endpoint`)
- [basicauth extension](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/extension/basicauthextension#configuration) (see `inline`)

Why should environment variables work differently from other providers?

Reply:
The `${THING_A}${THING_B}` pattern is also interesting. Right now, IIUC, we interpret this as "get the thing with URI `THING_A}${THING_B`", which also seems counter to what I would expect to happen (I guess this relates to #5706 too).

Reply:
To make progress, I disabled this feature from the Resolver. For the moment no way to enable this, but will send another PR that addresses the remaining issues and use a featuregate to enable this maybe.

Reply:
Makes sense. Sorry to bring this up so late, I hadn't realized until now